### ISSUE TITLE
Deprecate `is_executable`

### ISSUE THREAD

**Issue by traveltissues** [bazelbuild/remote-apis#99]
If #91 is merged I think `is_executable` can be deprecated in v3 and this can be handled via properties.

**Comment by ola-rozenfeld** [comment]
Pros of deprecating: `is_executable` logically belongs with the rest of the node properties.
Cons of deprecating:
* currently only takes two bytes to specify `is_executable`, using node properties will take at least 10 bytes.
* migration will be annoying

**Comment by edbaunton** [comment]
I would rather we didn't deprecate this property. This would make #91 a non-optional extension to the API. Currently, I am not keen on the current design because of (at least) the issues mentioned above. Deprecating `is_executable` would force support for these properties.

Admittedly I did not comment on the original #91 so please excuse: but I haven't heard of many compelling use cases other than the BuildStream incremental build use cases.

**Comment by edbaunton** [comment]
I am curious as to the rationale for using a Key/Value pair with a versioned lexicon over extending the API for each concrete property we think is necessary to support?

I understand not each property may make sense on each platform that is supported, however, that doesn't seem like it would be a blocker; the same way the current design would ignore unsupported keys. The capabilities response must report the superset in any case.

**Comment by juergbi** [comment]
> I am curious as to the rationale for using a Key/Value pair with a versioned lexicon over extending the API for each concrete property we think is necessary to support?

I see the following advantages in the key/value approach:
- Less friction for experiments and niche use cases. While protobufs are extensible, they don't allow multiple parties to independently add fields without conflicts.
- There may be code that can work with arbitrary node properties. E.g., code to merge two directory trees can simply copy node properties as long as there are no file conflicts.
- The same string keys used in `NodeProperty` can also be used for `output_node_properties` in `Action` and `supported_node_properties` in `ExecutionCapabilities`. Otherwise we'd need some kind of mapping and handle that in every code base that wants to support node properties.

Overall the reasoning is similar to platform properties. It's definitely less efficient, however, I currently don't expect a large number of actions to use node properties in a significant way. E.g., while BuildStream will need `MTime` for incremental builds, we don't intend to use `MTime` for the vast majority of builds, which are non-incremental. BuildStream may also support extended permission bits in the future, however, I also expect that we make this opt-in for builds that really need it.

I'm not against keeping `is_executable` as is, however, this issue is about deprecating `is_executable` in v3, not in v2. I'd suggest we reevaluate this when v3 is a bit closer and we have gained some experience with regards to node properties.

**Comment by sstriker** [comment]
I also tend to lean towards not deprecating is_executable from the efficiency standpoint.

> Less friction for experiments and niche use cases. While protobufs are extensible, they don't allow multiple parties to independently add fields without conflicts.

True.  Hoever, for these experiments to work you need to control both a client and a server.  There isn't much stopping from using an extended protobuf during experimentation.  Once the experiment proves successful getting the extensions integrated into API proper then makes sense.  If it isn't part of the API there won't be an incentive for servers or clients to implement the desired semantics.

> There may be code that can work with arbitrary node properties. E.g., code to merge two directory trees can simply copy node properties as long as there are no file conflicts.

True.

> The same string keys used in NodeProperty can also be used for output_node_properties in Action and supported_node_properties in ExecutionCapabilities. Otherwise we'd need some kind of mapping and handle that in every code base that wants to support node properties.

Yes.  However, if node properties are first class citizens in client API SDKs then I don't really see an issue here, considering that the number of server implementations will likely be more limited.

While I see the comparison with platform properties, there are likely few actions in comparison to files.  That is, for each action many node properties may need to be considered.

In summary: I like the idea of concrete properties, especially common ones.  We should take note of what they are and if prolific enough add them to API proper.  That may be slightly more expensive in terms of implementation, but that's a fair trade-off I'd say.  Given that we start off with is_executable and mtime, we may have jumped the gun with making this generic.

Given NodeProperty is a string to string map, I imagine server operators will want to eliminate the use of properties in favour of something more efficient.  For implementations like buildbox-casd this is likely less problematic, as communication with a client is local.

**Comment by ulfjack** [comment]
Given commit 7802003e00901b4e740fe0ebec1243c221e02ae2, is this discussion moot and we can move forward with the deprecation?

**Comment by sstriker** [comment]
@ulfjack Now that we have shifted to field-based node-properties, I think we should revisit.  I'll add it to the agenda for next meeting as part of the triage.
